src/v/cluster/namespaced_cache.h

```diff
@@ -0,0 +1,497 @@
+
+    size_t effective_min_slots_per_namespace() const {
+        return std::min(_max_size(), _min_slots_per_namespace());
+    }
+
+    void handle_max_size_change();
+
+    namespaces_t _namespaces;
+
+    size_t _size{0};
+    config::binding<size_t> _max_size;
+    config::binding<size_t> _min_slots_per_namespace;
+    PreEvictionHookT _pre_eviction_hook;
+    PostEvictionHookT _post_eviction_hook;
+};
+
+/**
+ * namespaced_cache definitions
+ */
+
+template<
+  typename EntryT,
+  typename NamespaceT,
+  safe_intrusive_list_hook EntryT::*HookPtr,
+  pre_eviction_hook<EntryT> PreEvictionHookT,
+  post_eviction_hook<EntryT> PostEvictionHookT,
+  namespace_cache<EntryT, HookPtr, PreEvictionHookT, PostEvictionHookT> CacheT>
+bool namespaced_cache<
+  EntryT,
+  NamespaceT,
+  HookPtr,
+  PreEvictionHookT,
+  PostEvictionHookT,
+  CacheT>::evict(namespaces_t::iterator hint) {
+    /**
+     * linear scan over all namespaces to find the one with the largest number
+     * entries in an LRU
+     */
+    auto max_it = std::max_element(
+      _namespaces.begin(),
+      _namespaces.end(),
+      [](const auto& lhs, const auto& rhs) {
+          return lhs.second->size() < rhs.second->size();
+      });
+
+    /**
+     * If hint is provided and the number of elements in a hint pointed
+     * namespace cache is equal to max element size evict from the hint
+     * namespace
+     */
+    if (
+      hint != _namespaces.end()
+      && max_it->second->size() == hint->second->size()) {
+        max_it = hint;
+    }
+
+    std::unique_ptr<CacheT>& t_state = max_it->second;
+    auto evicted = t_state->evict(_pre_eviction_hook, _post_eviction_hook);
+    if (evicted) {
+        --_size;
+    }
+    return evicted;
+}
+
+template<
+  typename EntryT,
+  typename NamespaceT,
+  safe_intrusive_list_hook EntryT::*HookPtr,
+  pre_eviction_hook<EntryT> PreEvictionHookT,
+  post_eviction_hook<EntryT> PostEvictionHookT,
+  namespace_cache<EntryT, HookPtr, PreEvictionHookT, PostEvictionHookT> CacheT>
+void namespaced_cache<
+  EntryT,
+  NamespaceT,
+  HookPtr,
+  PreEvictionHookT,
+  PostEvictionHookT,
+  CacheT>::insert(const NamespaceT& ns, EntryT& entry) {
+    if (
+      _namespaces.size() >= namespace_capacity() && !_namespaces.contains(ns)) {
+        throw cache_full_error(ssx::sformat(
+          "maximum number of namespaces reached. Min number of entries per "
+          "namespace: {}, max cache capacity: {}, current namespaces: {}",
+          effective_min_slots_per_namespace(),
+          _max_size(),
+          _namespaces.size()));
+    }
+    auto [it, _] = _namespaces.try_emplace(ns, std::make_unique<CacheT>());
+
+    /**
+     * It is very unlikely that this loop will take more than few iterations to
+     * find the eviction spot. Therefore we do not yield here.
+     *
+     * An auxiliary stop condition allow the cache to temporarily grow beyond
+     * the max size if any of the elements can be evicted.
```

Comment:
should we rather throw an error in such a case? In some hypothetical edge case (eg: all transactions are open on all producers, so none of them can evict), we just keep inserting with this logic in place.

Reply:
makes sense, i've changed the logic here